src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash128.cs

```diff
@@ -0,0 +1,460 @@
+            }
+
+            if (bound >= 2)
+            {
+                Mix32Bytes(ref accLow, ref accHigh, source + (32 * 5), source + (32 * 5) + 16, DefaultSecret3UInt64_4, DefaultSecret3UInt64_5, DefaultSecret3UInt64_6, DefaultSecret3UInt64_7, seed);
+            }
+
+            if (bound == 3)
+            {
+                Mix32Bytes(ref accLow, ref accHigh, source + (32 * 6), source + (32 * 6) + 16, DefaultSecret3UInt64_8, DefaultSecret3UInt64_9, DefaultSecret3UInt64_10, DefaultSecret3UInt64_11, seed);
+            }
+
+            Mix32Bytes(ref accLow, ref accHigh, source + length - 16, source + length - 32, 0x4F0BC7C7BBDCF93F, 0x59B4CD4BE0518A1D, 0x7378D9C97E9FC831, 0xEBD33483ACC5EA64, 0 - seed);
+
+            return AvalancheHash(accLow, accHigh, length, seed);
+        }
+
+        private static Hash128 HashLengthOver240(byte* source, uint length, ulong seed)
+        {
+            Debug.Assert(length > 240);
+
+            fixed (byte* defaultSecret = &MemoryMarshal.GetReference(DefaultSecret))
+            {
+                byte* secret = defaultSecret;
+                if (seed != 0)
+                {
+                    byte* customSecret = stackalloc byte[SecretLengthBytes];
+                    DeriveSecretFromSeed(customSecret, seed);
+                    secret = customSecret;
+                }
+
+                ulong* accumulators = stackalloc ulong[AccumulatorCount];
+                InitializeAccumulators(accumulators);
+
+                HashInternalLoop(accumulators, source, length, secret);
+
+                Hash128 h128;
+                h128.Low64 = MergeAccumulators(accumulators, secret + SecretMergeAccsStartBytes, length * Prime64_1);
+                h128.High64 = MergeAccumulators(accumulators, secret + SecretLengthBytes - AccumulatorCount * sizeof(ulong) - SecretMergeAccsStartBytes, ~(length * Prime64_2));
+                return h128;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Hash128 AvalancheHash(ulong accLow, ulong accHigh, uint length, ulong seed)
+        {
+            ulong h128Low = accLow + accHigh;
+            ulong h128High = (accLow * Prime64_1)
+                          + (accHigh * Prime64_4)
+                          + ((length - seed) * Prime64_2);
+            h128Low = Avalanche(h128Low);
+            h128High = 0ul - Avalanche(h128High);
+            return new Hash128(h128Low, h128High);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong XorShift64(ulong v64, int shift)
+        {
+            Debug.Assert(0 <= shift && shift < 64);
+            return v64 ^ (v64 >> shift);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong Multiply32To64(uint v1, uint v2) => (ulong)v1 * v2;
```

Comment:
We have a variant of this in XxHashShared.  Can they be deduped?

Reply:
This one is slightly different. The one in XxHashShared multiplies 2 ulong, while this one multiplies 2 uint and is only used by XxHash128.

Reply:
> The one in XxHashShared multiplies 2 ulong

But it first casts them to uint, and it's only used in one place.

Reply:
> But it first casts them to uint, and it's only used in one place.

Ah, I haven't checked the other. Indeed if that's the case.

Reply:
Fixed by commit 620ef11